internal/users/manager.go

```diff
@@ -123,33 +149,24 @@ func (m *Manager) UpdateUser(u types.UserInfo) (err error) {
 	}
 
 	var uid uint32
-
-	// Prevent a TOCTOU race condition between the check for existence in our database and the registration of the
-	// temporary user/group records. This does not prevent a race condition where a user is created by some other NSS
-	// source, but that is handled in the temporaryRecords.RegisterUser and temporaryRecords.RegisterGroup functions.
-	m.updateUserMu.Lock()
-	defer m.updateUserMu.Unlock()
+	var isNewUser bool
 
 	// Check if the user already exists in the database
 	oldUser, err := m.db.UserByName(u.Name)
 	if err != nil && !errors.Is(err, db.NoDataFoundError{}) {
 		return fmt.Errorf("could not get user %q: %w", u.Name, err)
 	}
 	if errors.Is(err, db.NoDataFoundError{}) {
-		// Check if the user exists on the system
-		existingUser, err := user.Lookup(u.Name)
-		var unknownUserErr user.UnknownUserError
-		if !errors.As(err, &unknownUserErr) {
-			log.Errorf(context.Background(), "User already exists on the system: %+v", existingUser)
-			return fmt.Errorf("user %q already exists on the system (but not in this authd instance)", u.Name)
+		isNewUser = true
+		localEntries, unlockEntries, err := localentries.NewWithLock()
+		if err != nil {
+			return err
 		}
+		defer func() { err = errors.Join(err, unlockEntries()) }()
+		records := m.temporaryRecords.LockForChanges(localEntries)
 
-		// The user does not exist, so we generate a unique UID for it. To avoid that a user with the same UID is
-		// created by some other NSS source, this also registers a temporary user in our NSS handler. We remove that
-		// temporary user before returning from this function, at which point the user is added to the database (so we
-		// don't need the temporary user anymore to keep the UID unique).
 		var cleanup func()
-		uid, cleanup, err = m.temporaryRecords.RegisterUser(u.Name)
+		uid, cleanup, err = records.RegisterUser(u.Name)
```

Comment:
We also don't need any temporary records here anymore, we can instead take a lock, choose a unique UID, store it in the database, and then release the lock. Yes, that means that `UpdateUser` functions won't run in parallel anymore, but IMO the reduction of complexity outweighs that.

Reply:
True, the changes to improve the ID generation also introduce quite a lot lines of code. However, those would add up with the code I was able to drop by taking a lock.